### PR TITLE
Use pybind11

### PR DESCRIPTION
**Pull request by joakim-hove**
OK - this should be reasonably ready now.

This PR is hit by the GitHub-commit-reorder syndrom, the: "Added pybind11 as git submodule" is the first commit in the tree.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -36,12 +37,17 @@ before_script:
...
+  - make install
+  - popd
   - popd
   - mkdir build
   - cd build
 
 script:
-  - cmake -DCMAKE_PREFIX_PATH=$INSTALL_ROOT -DUSE_RPATH=ON -DCMAKE_INSTALL_PREFIX=$INSTALL_ROOT ..
+  - cmake -DCMAKE_PREFIX_PATH=$INSTALL_ROOT -DUSE_RPATH=ON -DCMAKE_INSTALL_PREFIX=$INSTALL_ROOT -DPYTHON_EXECUTABLE=`which python` ..
   - make install
```
Comment (ReedOnly) [inline .travis.yml]: Use only `make` to make Travis build finish succesfully

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,10 @@ enable_language( CXX )
 include( CheckCXXSourceCompiles )
 include( CTest )
 
+find_package(pybind11 REQUIRED)
```
Comment (ReedOnly) [inline CMakeLists.txt]: Use `git submodule` to add linking to pybind11

Comment (joakim-hove) [reply CMakeLists.txt]: Personally I would prefer to assume that `pybind11` is installed on the system in question, and just use the `find_package(pybind11 REQUIRED)`. I agree we can use `git submodule` for `pybind11` - but that would be a "general decision" - and not just to support Travis building.

File: tests/schedule.py [hunk 3]
```diff
@@ -24,13 +24,13 @@ def testContains(self):
...
-        self.assertEqual(dt.date(2015, 1, 1),   self.sch.start)
-        self.assertEqual(dt.date(2029, 12, 27), self.sch.end)
+        self.assertEqual(dt.date(2015, 1, 1),   self.sch.start.date())
+        self.assertEqual(dt.date(2029, 12, 27), self.sch.end.date())
 
     def testTimesteps(self):
         timesteps = self.sch.timesteps
         self.assertEqual(176, len(timesteps))
-        self.assertEqual(dt.date(2015, 12, 31), timesteps[7])
+        self.assertEqual(dt.date(2015, 12, 31), timesteps[7].date())
```
Comment (pgdr) [inline tests/schedule.py]: please see if you can do this without changing the API

Comment (joakim-hove) [reply tests/schedule.py]: We can probably do this without changing the API, but the API was in my opionon slightly broken on this particular point - the timepoints given by the `Schedule` object were in the form of `python datetime.date` instances, but the underlying datamodel has second resolution so python `datetime.datetime` seems more appropriate.

Comment (pgdr) [reply tests/schedule.py]: Sounds good, please do that.

**Comment by pgdr** [comment]
Please write commit messages in imperative form.

File: tests/schedule.py (hunk 3 above)
Comment (joakim-hove) [reply tests/schedule.py]: > please do that.

Please do what?

Comment (pgdr) [reply tests/schedule.py]: See if you can complete the refactoring without changing the tests?  Or isn't `timesteps[7]` supposed to be a date?

Comment (pgdr) [reply tests/schedule.py]: If not, then we should change the left-hand side to contain the hour/minute/seconds part.

Comment (joakim-hove) [reply tests/schedule.py]: > Or isn't timesteps[7] supposed to be a date?

No timestep is supposed to be full datetime, that was I why i considered the previous API to be broken. 

> then we should change the left-hand side to contain the hour/minute/seconds part.

I agree that is a better fix - will do that.

**Comment by joakim-hove** [comment]
> Please write commit messages in imperative form.

Done

File: tests/schedule.py (hunk 3 above)
Comment (joakim-hove) [reply tests/schedule.py]: > I agree that is a better fix - will do that.

Done. I let the the code changing the test be in a separate commit - so that it would stand out clearly for review; but I guess it should really be melded together with the main commit?

File: .travis.yml [hunk 4]
```diff
@@ -37,11 +41,14 @@ before_script:
...
   - popd
   - popd
-  - mkdir build
-  - cd build
+  - pip install --user -r requirements.txt
```
Comment (JensGM) [inline .travis.yml]: You should be able to install without --user, I think even without sudo

File: .travis.yml [hunk 5]
```diff
@@ -37,11 +41,14 @@ before_script:
...
 
 script:
+  - mkdir build
+  - pushd build
   - cmake -DCMAKE_PREFIX_PATH=$INSTALL_ROOT -DUSE_RPATH=ON -DCMAKE_INSTALL_PREFIX=$INSTALL_ROOT ..
   - make install
-  - export PYTHONPATH=$PYTHONPATH:$INSTALL_ROOT/lib/python${PYTHON_VERSION}/site-packages
+  - export PYTHONPATH=$PYTHONPATH:$INSTALL_ROOT/lib/python${PYENV_VERSION}/site-packages
   - ctest --output-on-failure
+  - popd
```
Comment (JensGM) [inline .travis.yml]: Missing linebreak

File: .travis.yml [hunk 6]
```diff
@@ -37,11 +41,14 @@ before_script:
...
+  - pip install --user -r requirements.txt
+
 
 script:
+  - mkdir build
+  - pushd build
   - cmake -DCMAKE_PREFIX_PATH=$INSTALL_ROOT -DUSE_RPATH=ON -DCMAKE_INSTALL_PREFIX=$INSTALL_ROOT ..
   - make install
-  - export PYTHONPATH=$PYTHONPATH:$INSTALL_ROOT/lib/python${PYTHON_VERSION}/site-packages
+  - export PYTHONPATH=$PYTHONPATH:$INSTALL_ROOT/lib/python${PYENV_VERSION}/site-packages
```
Comment (JensGM) [inline .travis.yml]: Is setting the python path necessary?

File: sunbeam/eclipse_state.cpp [hunk 7]
```diff
@@ -83,18 +85,18 @@ namespace {
...
-    py::class_< EclipseState >( "EclipseState", py::no_init )
-        .add_property( "title", &EclipseState::getTitle )
-        .def( "_schedule",      &EclipseState::getSchedule,     ref() )
-        .def( "_props",         &EclipseState::get3DProperties, ref() )
-        .def( "_grid",          &EclipseState::getInputGrid,    ref() )
-        .def( "_cfg",           &EclipseState::cfg,             ref() )
-        .def( "_tables",        &EclipseState::getTableManager, ref() )
+    py::class_< EclipseState >( module, "EclipseState" )
+        .def_property_readonly( "title", &EclipseState::getTitle )
+        .def( "_schedule",      &EclipseState::getSchedule)
```
Comment (JensGM) [inline sunbeam/eclipse_state.cpp]: These seem to return copies now, rather than references. Is that intentional?

File: sunbeam/eclipse_grid.cpp [hunk 8]
```diff
@@ -13,38 +13,43 @@ namespace {
...
+    py::class_< EclipseGrid >( module, "EclipseGrid")
         .def( "_getXYZ",        &getXYZ )
         .def( "nactive",        &getNumActive )
         .def( "cartesianSize",  &getCartesianSize )
         .def( "globalIndex",    &getGlobalIndex )
         .def( "getIJK",         &getIJK )
         .def( "_cellVolume1G",  &cellVolume1G)
-        .def( "_cellVolume3",   &cellVolume3)
-        ;
+        .def( "_cellVolume3",   &cellVolume3);
```
Comment (JensGM) [inline sunbeam/eclipse_grid.cpp]: Not having the semi colon on a separate line makes these slightly more annoying to modify/organize.

File: tests/spe3/SPE3CASE1.DATA [hunk 9]
```diff
@@ -468,7 +468,7 @@ WCONINJE
 
 TSTEP
 -- first year:
-7*52.14285714
+6*52 53
```
Comment (JensGM) [inline tests/spe3/SPE3CASE1.DATA]: Is this related to the timezones?

File: .travis.yml (hunk 4 above)
Comment (joakim-hove) [reply .travis.yml]: I fought *a lot* with Travis; at one stage I certainly needed the `--user`- but maybe it can be removed now- will try.

File: .travis.yml (hunk 6 above)
Comment (joakim-hove) [reply .travis.yml]: Well this test runs on the installed version of sunbeam - which I think is a good thing - and then the `PYTHONPATH` manipulations are required.

File: sunbeam/eclipse_grid.cpp (hunk 8 above)
Comment (joakim-hove) [reply sunbeam/eclipse_grid.cpp]: And with the `;` on separate lines they look dang ugly. Anyway - they were on separate lines previosuly, I will restore that.

File: sunbeam/eclipse_state.cpp (hunk 7 above)
Comment (joakim-hove) [reply sunbeam/eclipse_state.cpp]: To be quite honest - going through the memory model with respect to references / copies / parents and what not was still on my list; that also includes running Valgrind.

File: tests/spe3/SPE3CASE1.DATA (hunk 9 above)
Comment (joakim-hove) [reply tests/spe3/SPE3CASE1.DATA]: The original expression:
```
7*52.14285714
```
correponds to dividing the year in seven identical parts. The problem is that when this is converted to seconds in opm-parser we get:
```
double seconds = 52.14285714 * 86400 = 4505142.8569 
```
and when that is converted to `std::time_t` we loose the fractional part in each conversion, and after seven timesteps have elapsed we are six seconds short of a year. I opted to change this to timesteps which can be perfectly represented as seconds - then we can tests like this:
```
self.assertEqual( datetime.datetime( 2016,1,1), timestep[7])
```
One advantage of this is that we will be future proof if/when opm-parser supports sub-second resolution in the timesteps. Alternatively we can keep things as they are - but then the test must be like:
```
self.assertEqual( datetime.date( 2015, 12, 31, 23, 59, 54), timestep[7])
```
I have a slight preference for the change I have made, but will *not' fight for it.

File: .travis.yml (hunk 4 above)
Comment (JensGM) [reply .travis.yml]: You don't need sudo in if the language is set to python. But since this is set to c++ (?), you might need --user, or sudo.

File: .travis.yml (hunk 6 above)
Comment (JensGM) [reply .travis.yml]: Alright

**Comment by joakim-hove** [comment]
> [...] did this solve the memory issues?

Yes - now they are solved; all good in Valgrind

File: sunbeam/eclipse_state.cpp (hunk 7 above)
Comment (joakim-hove) [reply sunbeam/eclipse_state.cpp]: OK - this has been fixed.

**Comment by joakim-hove** [comment]
OK - with the exception of what to do with the time testing I now consider this ready for final review/merging?

File: tests/spe3/SPE3CASE1.DATA (hunk 9 above)
Comment (pgdr) [reply tests/spe3/SPE3CASE1.DATA]: > I have a slight preference for the change I have made, but will not fight for it.

But then it's not SPE3CASE1 anymore, or?

Comment (joakim-hove) [reply tests/spe3/SPE3CASE1.DATA]: This is not imporant for me - I have presented some arguments why I think we should modify the SPE3 test deck, but if the general opinion on the matter is that the SPE3 test deck should be left unmodified I will revert the changes.

> But then it's not SPE3CASE1 anymore, or?

The offical SPEx cases are described in papers from Society of Petroleum Engineers (SPE), there exists no official instances as input decks. The SPE3 case we started with were manually created based on the original papers by a summer student in 2015(?). So - the changes suggested in this PR does not dramatically change the canonical SPE3 state of the case - and anyway that does not really matter that much?

But as I said; this is not important to me. I will not argue the case any further, but happily revert the changes if told to do so.

Comment (pgdr) [reply tests/spe3/SPE3CASE1.DATA]: Okay, I agree to the change of SPE3, but I would like a comment on that line
```
TSTEP
-- first year:
6*52 53 -- was originally "7*52.14285714", changed in sunbeam#58
```

File: .travis.yml (hunk 5 above)
Comment (joakim-hove) [reply .travis.yml]: Fixed
